### PR TITLE
all: move main transaction pool into a subpool

### PR DESCRIPTION
**Pull request by karalabe**
This PR is a refactor to move the `txpool.TxPool` implementation to `legacypool.LegacyPool`, which will be a sibling to the upcoming `blobpool.BlobPool`. The integration of the 2 pools will be lightyears simpler if the two are siblings and not parent-child relationship. This abstraction also allows us to move the local txs into their own pool and further simplify the main txpool (legacypool from now on).

The reason I named it legacypool instead of something more permanent is because I think we should rethink and reimplement the txpool to use some persistency soon rather than later, so might as well start that idea now.

### CODE REVIEW

File: core/txpool/subpool.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	// start up.
+	//
+	// These should not be passed as a constructor argument - nor should the pools
+	// start by themselves - in order to keep multiple subpools in lockstep with
+	// one another.
+	Init(gasTip *big.Int, head *types.Header) error
+
+	// Close terminates any background processing threads and releases any held
+	// resources.
+	Close() error
```
Comment (holiman) [inline core/txpool/subpool.go:60]: legacy pool doesn't ever return an error (only `nil`). You might skip this returnvalue and simplify the txpoo.go Close method err-handling a bit (unless you foresee the blobpool to have a reason to return errors on closing)

File: core/txpool/txpool.go [hunk 2]
```diff
@@ -145,1705 +36,296 @@ const (
...
-		if removed, _ := future.Remove(tx); removed {
-			// Reduce the queued counter
-			queuedGauge.Dec(1)
-		}
-		if future.Empty() {
-			delete(pool.queue, addr)
-			delete(pool.beats, addr)
+// Nonce returns the next nonce of an account, with all transactions executable
+// by the pool already applied on top.
+func (p *TxPool) Nonce(addr common.Address) uint64 {
```
Comment (holiman) [inline core/txpool/txpool.go:247]: Yeah, this is a bit wonky, api-wise. I guess in the future, the nonce tracking need to be somehow centralized. Works for now though

File: core/types/transaction_test.go [hunk 3]
```diff
@@ -281,7 +281,7 @@ func testTransactionPriceNonceSort(t *testing.T, baseFee *big.Int) {
 	signer := LatestSignerForChainID(common.Big1)
 
 	// Generate a batch of transactions with overlapping values, but shifted nonces
-	groups := map[common.Address]Transactions{}
+	groups := map[common.Address][]*Transaction{}
```
Comment (holiman) [inline core/types/transaction_test.go:284]: Why did you make these changes?

File: core/txpool/subpool.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto/kzg4844"
+	"github.com/ethereum/go-ethereum/event"
+)
+
+// Transaction is a helper struct to group together a canonical transaction with
+// satellite data items that are needed by the pool but are not part of the chain.
+type Transaction struct {
+	Tx *types.Transaction // Canonical transacrion
```
Comment (holiman) [inline core/txpool/subpool.go:32]:
```suggestion
	Tx *types.Transaction // Canonical transaction
```

File: event/multisub.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package event
+
+// JoinSubscriptions joins multiple subscriptions to be able to track them as
+// one entity and collectively cancel them of consume any errors from them.
+func JoinSubscriptions(subs ...Subscription) Subscription {
```
Comment (holiman) [inline event/multisub.go:21]: This is nice and neat, but you _could_ have used a much simpler one, right? You know how many subpools you have,  so you could have had one goroutines listening to all of them, instead of one routine each: 

```golang
			select {
				case err := <-sub1.Err():
					if err != nil {
						errc <- err
					}
				case err := <-sub2.Err():
					if err != nil {
						errc <- err
					}
...
```
Doesn't matter though, your solution is more generic

**Review by holiman — APPROVED** [review]
All in all, looks good to me

File: core/txpool/subpool.go (hunk 1 above)
Comment (karalabe) [reply core/txpool/subpool.go:60]: Myeah, it's just a PITA to add errors after the fact everywhere, so might as well do it properly. Also, the blobpool does return an error since it has a database that can fail (2 actually).

File: core/txpool/txpool.go (hunk 2 above)
Comment (karalabe) [reply core/txpool/txpool.go:247]: This is why I said that with the current API, every subpool needs to do a lot of crap. Unsure if it can be simplified though since they track different transactions in a different way.

File: core/types/transaction_test.go (hunk 3 above)
Comment (karalabe) [reply core/types/transaction_test.go:284]: The usage of types.Transactions was a bad ide / misguided. The reason that type was introduced is to have a derive sha method. But we started yoloing the code and using it as a shorthand for []*types.Transaction, which was not the goal. It's bad design to have 2 things mean the same thing in the code, hence why I try to be canonical and use the slice version.

File: event/multisub.go (hunk 5 above)
Comment (karalabe) [reply event/multisub.go:21]: The subpools can be added dynamically to the new txpool. Currently we have legacypool (and will have blobpool). We'll IMO add localpool, Marius wants to experiment with some disk backed pool. I can imagine enging up with a few controllable via a CLI flag (at least while testing implementations). It's simpler not to assume how many there will be and how they would be assembled.

File: core/txpool/txpool.go [hunk 6]
```diff
@@ -145,1705 +36,296 @@ const (
...
-		replaced, err := pool.add(tx, local)
-		errs[i] = err
-		if err == nil && !replaced {
-			dirty.addTx(tx)
+// Pending retrieves all currently processable transactions, grouped by origin
+// account and sorted by nonce.
+func (p *TxPool) Pending(enforceTips bool) map[common.Address][]*types.Transaction {
+	txs := make(map[common.Address][]*types.Transaction)
+	for _, subpool := range p.subpools {
+		for addr, set := range subpool.Pending(enforceTips) {
```
Comment (rjl493456442) [inline core/txpool/txpool.go:228]: Let's say we have [legacypool, blobpool]

the next nonce of account A is 5

in legacypool, it has tx(nonce=5), tx(nonce=6)
in blobpool, it has tx(nonce=7)

will tx pool somehow aggregate them together as a single executable batch,
or just tx(nonce=5), tx(nonce=6) are returned

File: go.mod [hunk 7]
```diff
@@ -2,14 +2,16 @@ module github.com/ethereum/go-ethereum
 
 go 1.19
 
+replace github.com/holiman/billy => github.com/karalabe/billy v0.0.0-20230331141121-732fa9109cff
```
Comment (rjl493456442) [inline go.mod:5]: Should we upstream the changes to `holiman/billy` and revert this change?

Comment (karalabe) [reply go.mod:5]: Ah yeah, the go.mod file should have been left out of this PR, thx.

File: core/txpool/txpool.go (hunk 6 above)
Comment (karalabe) [reply core/txpool/txpool.go:228]: Currently the invariant is that accounts are exclusive to pools. It is not yet enforced, but should be when blob pool lands. Specifically, if you have txs from an account in one pool, the others should reject it.

Comment (karalabe) [reply core/txpool/txpool.go:228]: Otherwise the whole point of multiple subpools becomes moot if they need to agree on state across each other and what to evict / how to track the same things.

File: core/txpool/txpool.go [hunk 8]
```diff
@@ -194,10 +194,17 @@ func (p *TxPool) Add(txs []*Transaction, local bool, sync bool) []error {
...
 	// happen that we receive merged batches, but better graceful than strange
 	// errors.
+	//
+	// We also need to track how the transactions were split across the subpools,
+	// so we can piece back the returned errors into the original order.
 	txsets := make([][]*Transaction, len(p.subpools))
 	splits := make([]int, len(txs))
 
 	for i, tx := range txs {
+		// Mark this tranaction belonging to no-subpool
```
Comment (holiman) [inline core/txpool/txpool.go:204]:
```suggestion
		// Mark this transaction belonging to no-subpool
```

**Comment by samlaf** [comment]
> I think we should rethink and reimplement the txpool to use some persistency soon rather than later

@karalabe has work on this started yet? And is there a design doc or issue explaining what you meant by this?
